crates/ide_assists/src/handlers/extract_variable.rs

```diff
@@ -146,6 +166,29 @@ fn valid_target_expr(node: SyntaxNode) -> Option<ast::Expr> {
     }
 }
 
+fn get_receiver_type(ctx: &AssistContext, expression: &ast::Expr) -> Option<hir::Type> {
+    let receiver = get_receiver(expression.to_owned())?;
+    Some(ctx.sema.type_of_expr(&receiver)?.original())
+}
+
+fn get_receiver(expression: ast::Expr) -> Option<ast::Expr> {
+    match expression {
+        ast::Expr::FieldExpr(field) if field.expr().is_some() => {
+            let nested_expression = &field.expr()?;
+            get_receiver(nested_expression.to_owned())
+        }
+        ast::Expr::PathExpr(_) => Some(expression),
```

Comment:
Is it correct to assume that in the expression `aaaa.bbbb.cccc.xxxx()`, (of type `FieldExpr`), if I keep going deeper into the nested receivers eventually I will end up at `aaaa` which is of type `PathExpr`? 

That's the type I seemed to get so I used that as the end condition of the recursion.

Reply:
in the general case yes, but this is not required, so this match should only have to branches, the one specializing on `ast::Expr::FieldExpr` and the other as the default just returning the current expression.

Reply:
Thanks! I've made that change as well as the others you remarked